### PR TITLE
[Performance] Improve Triton prefill attention kernel's performance 

### PR DESCRIPTION
**Pull request by Isotr0py**
## Purpose
Refine prefill attention kernel with FA2 implementation based on https://github.com/ModelTC/LightLLM/blob/2fbd2b8a7b180570129bd539bb34cbe0a5dbb22a/lightllm/common/basemodel/triton_kernel/att/prefill_att/context_flashattention_nopad.py#L287-L288
- In previous implementation, we use `float("-inf")` in attention mask, which needs additional invalid values mask for SWA. This PR replaces it with extremely large values $-10^8$ to overflow.
- Pre-scale `sm_scale` with $\frac{1}{ln(2)}$, then we can fuse `triton.math.exp` with faster `triton.math.exp2`

## Test Plan
Benchmark with https://github.com/noooop/snippet/blob/main/benchmarks/triton_attention/v1_offline_triton.py

## Test Result
**Main branch**
```
Batchsize 8, Input_len 2048 Throughput: 33.9996 requests/s, 69631.2040 tokens/s, Latency 113.12 ms, n_step 26
Batchsize 16, Input_len 2048 Throughput: 34.4612 requests/s, 70576.6036 tokens/s, Latency 207.27 ms, n_step 14
Batchsize 32, Input_len 2048 Throughput: 34.8346 requests/s, 71341.2651 tokens/s, Latency 358.84 ms, n_step 8
Batchsize 64, Input_len 2048 Throughput: 34.8565 requests/s, 71386.2010 tokens/s, Latency 478.15 ms, n_step 6
Batchsize 128, Input_len 2048 Throughput: 34.7642 requests/s, 71197.0056 tokens/s, Latency 719.13 ms, n_step 4
```

**Main branch + exp2 fusion**
```
Batchsize 8, Input_len 2048 Throughput: 34.4454 requests/s, 70544.1212 tokens/s, Latency 111.66 ms, n_step 26
Batchsize 16, Input_len 2048 Throughput: 34.7209 requests/s, 71108.4363 tokens/s, Latency 205.72 ms, n_step 14
Batchsize 32, Input_len 2048 Throughput: 35.1655 requests/s, 72018.9614 tokens/s, Latency 355.46 ms, n_step 8
Batchsize 64, Input_len 2048 Throughput: 35.2775 requests/s, 72248.3425 tokens/s, Latency 472.44 ms, n_step 6
Batchsize 128, Input_len 2048 Throughput: 34.9167 requests/s, 71509.3940 tokens/s, Latency 715.99 ms, n_step 4
```

**PR**
```
Batchsize 8, Input_len 2048 Throughput: 34.9690 requests/s, 71616.4340 tokens/s, Latency 109.99 ms, n_step 26
Batchsize 16, Input_len 2048 Throughput: 35.0962 requests/s, 71876.9206 tokens/s, Latency 203.52 ms, n_step 14
Batchsize 32, Input_len 2048 Throughput: 35.5592 requests/s, 72825.2077 tokens/s, Latency 351.53 ms, n_step 8
Batchsize 64, Input_len 2048 Throughput: 35.5579 requests/s, 72822.6684 tokens/s, Latency 468.72 ms, n_step 6
Batchsize 128, Input_len 2048 Throughput: 35.2918 requests/s, 72277.6372 tokens/s, Latency 708.38 ms, n_step 4
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Review by noooop — APPROVED** [review]

File: tests/models/language/pooling/test_token_classification.py [hunk 1]
```diff
@@ -86,7 +86,7 @@ def test_modernbert_models(
     for hf_output, vllm_output in zip(hf_outputs, vllm_outputs):
         hf_output = hf_output.detach().clone().cpu().float()
         vllm_output = vllm_output.detach().clone().cpu().float()
-        torch.testing.assert_close(hf_output, vllm_output, atol=1.2e-2, rtol=1e-3)
+        torch.testing.assert_close(hf_output, vllm_output, atol=3.2e-2, rtol=1e-3)
```
Comment (AndreasKaratzas) [inline tests/models/language/pooling/test_token_classification.py:89]: We were already seeing inaccuracies between HF and vLLM on ROCm, and we were thinking of fixing the accuracy. Were there any accuracy evaluations of this kernel especially for encoder-only self-attention models?

https://buildkite.com/vllm/amd-ci/builds/2766/steps/canvas?jid=019bb9da-1de2-4d4b-a7d5-f9bb8911e525

**Review by AndreasKaratzas — COMMENTED** [review]
@Isotr0py May I ask why there was this threshold increment? Was there a correctness issue?

File: tests/models/language/pooling/test_token_classification.py (hunk 1 above)
Comment (noooop) [reply tests/models/language/pooling/test_token_classification.py:89]: Please adjust the threshold to make this test pass. 

This test is only intended to verify that the model has been loaded correctly.

disham993/electrical-ner-ModernBERT-base is a randomly initialized model, not one that has been trained or fine-tuned, so numerical precision issues may arise.

Comment (AndreasKaratzas) [reply tests/models/language/pooling/test_token_classification.py:89]: Oh, I see. Thanks for that input :) Shall we be marking such models in CI such that we know when to care about accuracy and when not to care about accuracy? Or maybe use models that are deterministically initialized to ensure a deterministic CI?

Btw, threshold was increased so test will pass. It was lower before and sometimes randomly the test would fail.

Comment (AndreasKaratzas) [reply tests/models/language/pooling/test_token_classification.py:89]: @noooop Why is it compared with HF model then if we don't care about accuracy?

Comment (noooop) [reply tests/models/language/pooling/test_token_classification.py:89]: I didn't expect the gap to be this large.

Comment (AndreasKaratzas) [reply tests/models/language/pooling/test_token_classification.py:89]: Is there any model that you would recommend in place of modernbert? I can check and possibly push a PR if you've got a replacement to recommend.

Comment (noooop) [reply tests/models/language/pooling/test_token_classification.py:89]: To my knowledge, there is no more suitable repository on Hugging Face to use for testing.

Comment (AndreasKaratzas) [reply tests/models/language/pooling/test_token_classification.py:89]: Mm. I can add a comment there then, and a flaky retry decorator in hopes that the TG will have a near 100% pass rate. I'll cc you in an upcoming PR for that. In the meantime, we can think of a more elegant solution.
